javatests/arcs/core/data/expression/TypeCheckerTest.kt

```diff
@@ -0,0 +1,479 @@
+    )
+
+    checkTypeIs(true.asExpr() and true.asExpr(), BooleanType)
+    checkTypeIs(true.asExpr() or true.asExpr(), BooleanType)
+
+    val evaluator = checkTypeIs(1.asExpr() ifNull "hello".asExpr(), UnionType(IntType, TextType))
+    assertThat(evaluator.warnings).containsExactly("1 ?: \"hello\": 1 is always not null.")
+
+    val evaluator2 = checkTypeIs(nullExpr() ifNull "hello".asExpr(), UnionType(NullType, TextType))
+    assertThat(evaluator2.warnings).isEmpty()
+  }
+
+  @Test
+  fun check_field_ops() {
+    val emptyScope = MapScope<InferredType>("root", mapOf())
+    val aScope = ScopeType(emptyScope.builder().set("a", TextType).build())
+    val scope = emptyScope.builder().apply {
+      set("n", UnionType(aScope, NullType))
+      set("x", IntType)
+      set("y", DoubleType)
+      set("z", aScope)
+    }.build()
+    checkTypeIs(num("x"), IntType, scope)
+    checkTypeIs(num("y"), DoubleType, scope)
+    checkTypeIs(scope("z").get<InferredType>("a"), TextType, scope)
+    checkTypeIs(
+      scope("n").get<InferredType>("a"),
+      TextType,
+      scope,
+      errors = listOf("Field 'a` in n.a potentially looked up on null scope, use ?. operator.")
+    )
+  }
+
+  @Test
+  fun check_query_ops() {
+    val emptyScope = MapScope<InferredType>("root", mapOf())
+    val scope = emptyScope.builder().apply {
+      set("x", IntType)
+      set("y", DoubleType)
+    }.build()
+    checkTypeIs(query<Number>("x"), IntType, emptyScope, scope)
+    checkTypeIs(query<Number>("y"), DoubleType, emptyScope, scope)
+  }
+
+  @Test
+  fun check_from_expressions() {
+    val emptyScope = MapScope<InferredType>("root", mapOf())
+    val people = SeqType(ScopeType(emptyScope.builder().set("name", TextType).build()))
+    val scope =
+      emptyScope.builder().apply {
+        set("numbers", SeqType(IntType))
+        set("people", people)
+        set("company", SeqType(ScopeType(emptyScope.builder().set("people", people).build())))
+      }.build()
+
+    checkTypeIs(
+      PaxelParser.parse("from p in numbers select p"),
+      SeqType(IntType),
+      scope
+    )
+
+    checkTypeIs(
+      PaxelParser.parse("from p in people select p.name"),
+      SeqType(TextType),
+      scope
+    )
+
+    checkTypeIs(
+      PaxelParser.parse("from c in company from p in c.people select p.name"),
+      SeqType(TextType),
+      scope
+    )
+  }
+
+  @Test
+  fun check_where_expression() {
+    val emptyScope = MapScope<InferredType>("root", mapOf())
+    val scope = emptyScope.builder().apply {
+      set("numbers", SeqType(IntType))
+      set("x", TextType)
+    }.build()
+
+    checkTypeIs(
+      PaxelParser.parse("from p in numbers where p < 10 select p"),
+      SeqType(IntType),
+      scope
+    )
+
+    checkTypeIs(
+      PaxelParser.parse("from p in numbers where 2 + 3 select p"),
+      SeqType(IntType),
+      scope,
+      errors = listOf("from p in numbers\nwhere 2.0 + 3.0 must evaluate to a boolean type" +
```

Comment:
oh no, you caught a bug that "2" and "3" are parsed as doubles! This means that doing any arithmetic with constants immediately pushes you to doubles, which is not ideal... One could write 2i + 3i, but that's a bit weird, support for imaginary numbers so early on.. ;-)